File: HWs/old hws/2012f/HW4-5/text/printtopics.m
%----------------------------------------------------
% file:    topics.m
% purpose: print topics from topic model run
% usage:   matlab> topics
% inputs:  Nwt.txt, vocab.txt
% outputs: topics.txt
% author:  David Newman, newman@uci.edu
%----------------------------------------------------

[wid,tid,cnt] = textread('Nwt.txt','%d%d%d','headerlines',3);
Nwt = sparse(wid,tid,cnt);
[W,T] = size(Nwt);

beta = 0.01;
Nwt = Nwt + beta;

fac = 1 ./ sum(Nwt,1);
prob_w_given_t = Nwt * diag(fac);
  
prob_t = sum(Nwt,1);
prob_t = prob_t / sum(prob_t);

fid = fopen('topics2.txt','w');
[word] = textread('vocab.txt','%s');
assert(length(word)==W)

[xxsort,iisort] = sort(-prob_t);

for tt=1:T
  %t = iisort(tt);
  t = tt;
  prob = prob_w_given_t(:,t);
  [xsort,isort] = sort(-prob);
  %fprintf(fid,'[t%d] (prob=%.3f) ', t, prob_t(t));
  fprintf(fid,'[t%d] ', t);
  for w=1:12
    fprintf(fid,'%s ', word{isort(w)});
  end
  fprintf(fid,'...\n');
end

fclose(fid);
fprintf('wrote file topics2.txt\n');

File: HWs/old hws/2012f/HW4-5/text/printtopics.c
/*---------------------------------------------------
 * file:    topics.c
 * purpose: print topics
 * usage:   topics
 * inputs:  Nwt.txt, vocab.txt
 * author:  newman@uci.edu
 *-------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <assert.h>
#include "tlib.h"

/*==========================================
 * main
 *========================================== */
int main(int argc, char* argv[])
{
  int i, t, w, **Ntw, *Nt, dum;
  char **word;
  FILE *fp;
  int *prob_w_given_t, *indx;

  int T; // number of topics
  int W; // number of unique words
  
  if (argc != 1) {
    fprintf(stderr, "usage: %s\n", argv[0]);
    exit(-1);
  }
  
  Ntw = read_sparse_trans("Nwt.txt",&T,&W); assert(T>0); assert(W>0);
  fp = fopen("vocab.txt","r");              assert(fp);
  word = (char**)malloc(W*sizeof(char*));   assert(word);
  for (w = 0; w < W; w++) {
    word[w] = (char*)malloc(100*sizeof(char));
    dum = fscanf(fp,"%s",word[w]);
  }
  fclose(fp);

  prob_w_given_t = ivec(W);
  indx           = ivec(W);
  Nt             = ivec(T);

  for (t = 0; t < T; t++) {
    Nt[t] = 0;
    for (w = 0; w < W; w++) Nt[t] += Ntw[t][w];
  }
  
  for (t = 0; t < T; t++) {
    
    for (w = 0; w < W; w++) prob_w_given_t[w] = Ntw[t][w];
    isort(W, prob_w_given_t, -1, indx);
    
    printf("[t%d] ", t+1);
    for (i = 0; i < 8; i++) printf("%s ", word[indx[i]]);
    printf("...\n");
    
  }
  
  return 0;
}
